File: stream deck anti gravity/client-android/app/src/main/java/com/antigravity/streamdeck/data/model/Protocol.kt
package com.antigravity.streamdeck.data.model

import com.google.gson.annotations.SerializedName

data class WebSocketMessage<T>(
    @SerializedName("event") val event: String,
    @SerializedName("payload") val payload: T
)

object EventTypes {
    const val PRESS_BUTTON = "PRESS_BUTTON"
    const val RELEASE_BUTTON = "RELEASE_BUTTON"
    const val SWITCH_PAGE = "SWITCH_PAGE"
    const val SWITCH_PROFILE = "SWITCH_PROFILE"
    const val OPEN_FOLDER = "OPEN_FOLDER"
    const val GO_BACK_FOLDER = "GO_BACK_FOLDER"
    const val PING = "PING"
    const val PONG = "PONG"
    const val GRID_SYNC = "GRID_SYNC"
    const val BUTTON_STATE_CHANGE = "BUTTON_STATE_CHANGE"
    const val HARDWARE_SYNC = "HARDWARE_SYNC"
    const val SPOTIFY_SYNC = "SPOTIFY_SYNC"
}

data class PressButtonPayload(
    @SerializedName("profileId") val profileId: String,
    @SerializedName("pageIndex") val pageIndex: Int,
    @SerializedName("row") val row: Int,
    @SerializedName("col") val col: Int,
    @SerializedName("buttonId") val buttonId: String,
    @SerializedName("actionType") val actionType: String? = null
)

data class ButtonModel(
    @SerializedName("id") val id: String,
    @SerializedName("row") val row: Int,
    @SerializedName("col") val col: Int,
    @SerializedName("rowSpan") val rowSpan: Int? = 1,
    @SerializedName("colSpan") val colSpan: Int? = 1,
    @SerializedName("label") val label: String,
    @SerializedName("labelColor") val labelColor: String? = "#FFFFFF",
    @SerializedName("backgroundColor") val backgroundColor: String? = "#1E1E2E",
    @SerializedName("iconUrl") val iconUrl: String? = null,
    @SerializedName("state") val state: String? = "OFF",
    @SerializedName("states") val states: Map<String, ButtonStateModel>? = null,
    @SerializedName("actionType") val actionType: String? = null,
    @SerializedName("folderButtons") val folderButtons: List<ButtonModel>? = null
)

data class ButtonStateModel(
    @SerializedName("label") val label: String,
    @SerializedName("iconUrl") val iconUrl: String? = null,
    @SerializedName("backgroundColor") val backgroundColor: String? = "#1E1E2E"
)

data class GridSyncPayload(
    @SerializedName("activeProfileId") val activeProfileId: String,
    @SerializedName("activePageIndex") val activePageIndex: Int,
    @SerializedName("currentFolderId") val currentFolderId: String? = null,
    @SerializedName("folderPath") val folderPath: List<String>? = null,
    @SerializedName("gridConfig") val gridConfig: GridConfig,
    @SerializedName("buttons") val buttons: List<ButtonModel>
)

data class GridConfig(
    @SerializedName("rows") val rows: Int,
    @SerializedName("cols") val cols: Int
)

data class ButtonStateChangePayload(
    @SerializedName("buttonId") val buttonId: String,
    @SerializedName("newState") val newState: String
)

data class HardwareMetricsPayload(
    @SerializedName("cpuUsage") val cpuUsage: Int,
    @SerializedName("cpuTemp") val cpuTemp: Int,
    @SerializedName("ramUsage") val ramUsage: Int,
    @SerializedName("ramUsedGb") val ramUsedGb: String,
    @SerializedName("ramTotalGb") val ramTotalGb: String,
    @SerializedName("gpuUsage") val gpuUsage: Int
)

data class SpotifyTrackPayload(
    @SerializedName("title") val title: String,
    @SerializedName("artist") val artist: String,
    @SerializedName("isPlaying") val isPlaying: Boolean
)

File: stream deck anti gravity/server-windows/src/protocol.ts
export enum EventTypes {
  PRESS_BUTTON = 'PRESS_BUTTON',
  RELEASE_BUTTON = 'RELEASE_BUTTON',
  SWITCH_PAGE = 'SWITCH_PAGE',
  SWITCH_PROFILE = 'SWITCH_PROFILE',
  OPEN_FOLDER = 'OPEN_FOLDER',
  GO_BACK_FOLDER = 'GO_BACK_FOLDER',
  PING = 'PING',
  PONG = 'PONG',
  GRID_SYNC = 'GRID_SYNC',
  BUTTON_STATE_CHANGE = 'BUTTON_STATE_CHANGE',
  HARDWARE_SYNC = 'HARDWARE_SYNC',
  SPOTIFY_SYNC = 'SPOTIFY_SYNC'
}

export interface WebSocketMessage<T = any> {
  event: EventTypes | string;
  payload: T;
}

export interface PressButtonPayload {
  profileId: string;
  pageIndex: number;
  row: number;
  col: number;
  buttonId: string;
  actionType?: string;
}

export interface ButtonStateModel {
  label: string;
  iconUrl?: string;
  backgroundColor?: string;
}

export interface ButtonModel {
  id: string;
  row: number;
  col: number;
  rowSpan?: number;
  colSpan?: number;
  label: string;
  labelColor?: string;
  backgroundColor?: string;
  iconUrl?: string;
  state?: string;
  states?: Record<string, ButtonStateModel>;
  actionType?: string;
  actionPayload?: Record<string, any>;
  folderButtons?: ButtonModel[];
}

export interface GridConfig {
  rows: number;
  cols: number;
}

export interface GridSyncPayload {
  activeProfileId: string;
  activePageIndex: number;
  currentFolderId?: string;
  folderPath?: string[];
  gridConfig: GridConfig;
  buttons: ButtonModel[];
}

export interface ButtonStateChangePayload {
  buttonId: string;
  newState: string;
}

export interface HardwareMetricsPayload {
  cpuUsage: number;
  cpuTemp: number;
  ramUsage: number;
  ramUsedGb: string;
  ramTotalGb: string;
  gpuUsage: number;
}

export interface SpotifyTrackPayload {
  title: string;
  artist: string;
  isPlaying: boolean;
}
